### PR TITLE
[GPII-2955]: Monitoring GCP infra with Stackdriver Beta

### PR DESCRIPTION
**Pull request by natarajaya**
This enables Stackdriver Beta Monitoring / Logging APIs and updates resources manifests for GPII components (required for Kubernetes Dashboard to actually display resource metrics).
I am still working on alerting part of this, but it is covered by GPII-2967, which makes it out of scope of this PR.

### CODE REVIEW

**Comment by amatas** [comment]
LGTM

File: gcp/modules/couchdb/values.yaml [hunk 1]
```diff
@@ -21,3 +21,11 @@ affinity:
               values:
               - couchdb
         topologyKey: "kubernetes.io/hostname"
+
+resources:
+  requests:
+    cpu: 500m
+    memory: 256Mi
```
Comment (mrtyler) [inline gcp/modules/couchdb/values.yaml]: How did you decide on these limits?

File: shared/charts/gpii-flowmanager/values.yaml [hunk 2]
```diff
@@ -36,3 +36,13 @@ ingress:
...
 nameOverride: flowmanager
+
+## Optional resource requests and limits for deployment
+## ref: http://kubernetes.io/docs/user-guide/compute-resources/
+resources: {}
+  # requests:
+  #  cpu: 100m
+  #  memory: 128Mi
+  # limits:
+  #  cpu: 56
```
Comment (mrtyler) [inline shared/charts/gpii-flowmanager/values.yaml]: This value looks strange.

File: gcp/modules/nginx-ingress/values.yaml [hunk 3]
```diff
@@ -9,9 +9,23 @@ controller:
     maxReplicas: 4
     targetCPUUtilizationPercentage: 50
     targetMemoryUtilizationPercentage: 50
+  resources:
```
Comment (mrtyler) [inline gcp/modules/nginx-ingress/values.yaml]: Is there anything here that should go to the upstream chart?

File: shared/charts/gpii-flowmanager/values.yaml (hunk 2 above)
Comment (natarajaya) [reply shared/charts/gpii-flowmanager/values.yaml]: This is an example resource section I copied from some other chart. You don't think that our flowmanager deployment may require 56 CPUs and 256GB RAM one day? :)

File: gcp/modules/nginx-ingress/values.yaml (hunk 3 above)
Comment (natarajaya) [reply gcp/modules/nginx-ingress/values.yaml]: This is release values yaml, not a part of the chart.

File: gcp/modules/couchdb/values.yaml (hunk 1 above)
Comment (mrtyler) [reply gcp/modules/couchdb/values.yaml]: 😎

File: gcp/modules/nginx-ingress/values.yaml (hunk 3 above)
Comment (mrtyler) [reply gcp/modules/nginx-ingress/values.yaml]: Right, just checking.

File: shared/charts/gpii-flowmanager/values.yaml (hunk 2 above)
Comment (mrtyler) [reply shared/charts/gpii-flowmanager/values.yaml]: Ok. I didn't notice that the memory limit was GB.

File: gcp/modules/couchdb/values.yaml (hunk 1 above)
Comment (stepanstipl) [inline gcp/modules/couchdb/values.yaml]: I would strongly be in favor of setting the same values both for CPU & memory resources. This basically makes your cluster behave in a consistent manner (or more than this) and is kind of standard practice for this type of workloads.

File: gcp/modules/gke-cluster/main.tf [hunk 4]
```diff
@@ -12,10 +12,13 @@ module "gke_cluster" {
 
   initial_node_count = 1
   node_type          = "n1-standard-2"
-  kubernetes_version = "1.9.7-gke.6"
+  kubernetes_version = "1.10.7-gke.2"
```
Comment (stepanstipl) [inline gcp/modules/gke-cluster/main.tf]: Do we need 1.10 for stackdriver beta? Just that it's kind of unrelated otherwise and I would prefer to leave it out if it's not needed and perhaps do as separate PR.

File: gcp/modules/gpii-flowmanager/values.yaml [hunk 5]
```diff
@@ -15,3 +15,11 @@ ingress:
...
 
 datasource_hostname: "http://${couchdb_admin_username}:${couchdb_admin_password}@couchdb-svc-couchdb.gpii.svc.cluster.local"
+
+resources:
+  requests:
+    cpu: 500m
+    memory: 128Mi
+  limits:
+    cpu: "1"
+    memory: 256Mi
```
Comment (stepanstipl) [inline gcp/modules/gpii-flowmanager/values.yaml]: Same as above

File: gcp/modules/gpii-preferences/values.yaml [hunk 6]
```diff
@@ -15,3 +15,11 @@ ingress:
...
 
 datasource_hostname: "http://${couchdb_admin_username}:${couchdb_admin_password}@couchdb-svc-couchdb.gpii.svc.cluster.local"
+
+resources:
+  requests:
+    cpu: 500m
+    memory: 128Mi
+  limits:
+    cpu: "1"
+    memory: 256Mi
```
Comment (stepanstipl) [inline gcp/modules/gpii-preferences/values.yaml]: And once more ;)

File: gcp/modules/nginx-ingress/values.yaml [hunk 7]
```diff
@@ -9,9 +9,23 @@ controller:
...
 
 defaultBackend:
   replicaCount: 2
+  resources:
+    requests:
+      cpu: 10m
+      memory: 32Mi
+    limits:
+      cpu: 20m
+      memory: 64Mi
```
Comment (stepanstipl) [inline gcp/modules/nginx-ingress/values.yaml]: Same :)

File: shared/charts/gpii-flowmanager/README.md [hunk 8]
```diff
@@ -61,3 +61,4 @@ Parameter | Description | Default
 `image.repository` | container image repository | `gpii/universal`
 `image.checksum` | container image checksum | `sha256:8547f22ae8e86d7b4b09e10d9ec87b1605b47dc37904171c84555a55462f161e`
 `image.pullPolicy` | container image pullPolicy | `IfNotPresent`
+`resources` | optional resource requests and limits for deployment |
```
Comment (stepanstipl) [inline shared/charts/gpii-flowmanager/README.md]: Hm looks like the other options are described more in detail - should we describe the structure of resources as well?

File: shared/charts/gpii-flowmanager/templates/deployment.yaml [hunk 9]
```diff
@@ -8,7 +8,7 @@ spec:
   strategy:
     rollingUpdate:
       maxSurge: {{ .Values.rollingUpdate.maxSurge }}
-      maxUnavailable: {{ .Values.maxUnavailable }}
+      maxUnavailable: {{ .Values.rollingUpdate.maxUnavailable }}
```
Comment (stepanstipl) [inline shared/charts/gpii-flowmanager/templates/deployment.yaml]: I don't see this option mentioned in the README?

File: gcp/modules/gke-cluster/main.tf (hunk 4 above)
Comment (natarajaya) [reply gcp/modules/gke-cluster/main.tf]: Yes, 1.10 (is required)[https://cloud.google.com/monitoring/kubernetes-engine/installing] for Stackdriver Beta.

File: gcp/modules/couchdb/values.yaml (hunk 1 above)
Comment (natarajaya) [reply gcp/modules/couchdb/values.yaml]: @stepanstipl I don't understand this. Can you please elaborate?

File: shared/charts/gpii-flowmanager/README.md (hunk 8 above)
Comment (natarajaya) [reply shared/charts/gpii-flowmanager/README.md]: There is an [example resource section](https://github.com/natarajaya/gpii-infra/blob/monitoring/shared/charts/gpii-flowmanager/values.yaml#L40-L48) in chart values file.

**Comment by natarajaya** [comment]
@stepanstipl I don't have strong opinion on this topic. I checked number of examples from upstream charts ([1](https://github.com/helm/charts/blob/master/incubator/couchdb/values.yaml#L98-L104), [2](https://github.com/helm/charts/blob/master/stable/mongodb/values.yaml#L113-L119), [3](https://github.com/helm/charts/blob/master/stable/nginx-ingress/values.yaml#L329-L335), [4](https://github.com/helm/charts/blob/master/stable/phpmyadmin/values.yaml#L57-L62)) and did not find any consistency in there either :) However, I like your argument about more predictable scheduling. So, I am ready to go ahead with your proposition, just let me double check that nothing explodes in Stackdriver Kubernetes dashboard when resources and limits set to the same value.

File: gcp/modules/couchdb/values.yaml (hunk 1 above)
Comment (natarajaya) [reply gcp/modules/couchdb/values.yaml]: Oh nevermind, for a moment I thought you're proposing to set CPU and memory requests to the same value 😆

**Comment by mrtyler** [comment]
Still LGTM.

I agree with favoring predictability and simple behavior with an obvious failure mode over slightly more resiliency with a less clear failure mode.

**Comment by stepanstipl** [comment]
thanks @natarajaya, I'm very happy now 👍 

and thanks for fixing the `rollingUpdate` although it wasn't you who introduced it 👍 👍 

> Oh nevermind, for a moment I thought you're proposing to set CPU and memory requests to the same value 😆

🤣 would love to see how this would be done too!
